src/report_reader.cpp

```diff
@@ -275,79 +290,109 @@ std::string ReportReader<T>::Population::getDataUnits() const {
-    const auto update_layout = [&](const NodeID& node_id, const Selection::Range& range) {
-        concrete_node_ids.emplace_back(node_id);
-        result.node_ranges.emplace_back(range);
-
-        result.min_max_range.first = std::min(result.min_max_range.first, range.first);
-        result.min_max_range.second = std::max(result.min_max_range.second, range.second);
-        element_ids_count += (range.second - range.first);
-    };
-
     // Take all nodes if no selection is provided
     if (!node_ids) {
-        concrete_node_ids.reserve(node_ranges_.size());
-        result.node_ranges.reserve(node_ranges_.size());
-
-        for (const auto& node_range : node_ranges_) {
-            update_layout(node_range.first, node_range.second);
-        }
+        concrete_node_ids = node_ids_;
+        result.node_ranges = node_ranges_;
+        result.node_offsets = node_offsets_;
+        result.node_index = node_index_;
+        element_ids_count = node_offsets_.back();
     } else if (!node_ids->empty()) {
         const auto selected_node_ids = node_ids->flatten();
 
-        // Reserve space for all the requested IDs and shrink afterwards
-        concrete_node_ids.reserve(selected_node_ids.size());
-        result.node_ranges.reserve(selected_node_ids.size());
-
         for (const auto node_id : selected_node_ids) {
-            const auto it = node_ranges_.find(node_id);
-            if (it != node_ranges_.end()) {
-                update_layout(it->first, it->second);
+            const auto it = std::lower_bound(node_index_.begin(),
+                                             node_index_.end(),
+                                             node_id,
+                                             [&](const size_t i, const NodeID node_id) {
+                                                 return node_ids_[i] < node_id;
+                                             });
+
+            if (it != node_index_.end() && node_ids_[*it] == node_id) {
+                const auto& range = node_ranges_[*it];
+
+                concrete_node_ids.emplace_back(node_id);
+                result.node_ranges.emplace_back(range);
+                result.node_offsets.emplace_back(element_ids_count);
+                result.node_index.emplace_back(result.node_index.size());
+
+                element_ids_count += (range.second - range.first);
             }
         }
-
-        concrete_node_ids.shrink_to_fit();
-        result.node_ranges.shrink_to_fit();
     } else {
         // node_ids Selection exists, but is empty
     }
 
     // Extract the ElementIDs from the GIDs
-    if (!result.node_ranges.empty()) {
-        const auto min = result.min_max_range.first;
-        const auto max = result.min_max_range.second;
+    if (!concrete_node_ids.empty()) {
+        // Retrieve the maximum gap between blocks (defaults to 128MB == 8 x GPFS blocks)
+        const auto block_gap_limit_env = getenv("LIBSONATA_BLOCK_GAP_LIMIT");
```

Comment:
I agree that it would be a good idea; `getenv` is a funny function to deal with, the threadsafety is sus, and iirc, it triggers locale work (prob. only on first run?)  I try to avoid it, mainly due to previously being burned by it.

Reply:
Is there an alternative to `getenv`? I think I'd feel better if this was done statically at initialization time.

Reply:
Thank you guys for the comments. I have replaced the `getenv` with an extra parameter to the function and a boundary check, to prevent the users to set very small gaps and potentially force the code to trigger a hyperslab per GID (as in the original code changed in #174). The Python bindings are also updated and there are now CI tests for this as well.